src/freenet/clients/http/FirstTimeWizardNewToadlet.java

```diff
@@ -0,0 +1,295 @@
+                if (storageLimit < Node.MIN_STORE_SIZE * 5 / 4) // min store size + 10% for client cache + 10% for slashdot cache
+                    errors.put("storageLimitError", NodeL10n.getBase().getString("Node.invalidMinStoreSizeWithCaches"));
+                else if (storageLimit > (maxDatastoreSize = DatastoreUtil.maxDatastoreSize()))
+                    errors.put("storageLimitError",
+                            NodeL10n.getBase().getString("Node.invalidMaxStoreSize",
+                                    String.format("%.2f", (float) maxDatastoreSize / DatastoreUtil.oneGiB)));
+            } catch (NumberFormatException e) {
+                errors.put("storageLimitError",
+                        FirstTimeWizardNewToadlet.l10n("valid.number.prefix.storageLimit") + " " + e.getMessage());
+            }
+
+            if (!setPassword.isEmpty()) {
+                if (password.isEmpty())
+                    errors.put("passwordError", NodeL10n.getBase().getString("SecurityLevels.passwordNotZeroLength"));
+                if (password.length() > SecurityLevelsToadlet.MAX_PASSWORD_LENGTH)
+                    errors.put("passwordError", NodeL10n.getBase().getString("SecurityLevels.passwordTooLong"));
+                if (!password.equals(passwordConfirmation))
+                    errors.put("passwordError", NodeL10n.getBase().getString("SecurityLevels.passwordsDoNotMatch"));
+            }
+        }
+
+        private boolean isValid() {
+            return errors.isEmpty();
+        }
+
+        private void detectBandwidthLimit() {
+            final int KiB = 1024;
+            try {
+                BandwidthLimit detected =
+                        BandwidthManipulator.detectBandwidthLimits(core.node.ipDetector.getBandwidthIndicator());
+
+                // Detected limits reasonable; add half of both as recommended option.
+                downloadLimitDetected = Long.toString(detected.downBytes / 2 / KiB);
+                uploadLimitDetected = Long.toString(detected.upBytes / 2 / KiB);
+            } catch (Exception e) {
+                Logger.normal(this, e.getMessage(), e);
+            }
+        }
+
+        private Map<String, Object> toModel() {
+            HashMap<String, Object> model = new HashMap<>();
+            model.put("knowSomeone", knowSomeone.length() > 0 ? "checked" : "");
+            model.put("connectToStrangers", connectToStrangers.length() > 0 ? "checked" : "");
+            model.put("haveMonthlyLimit", haveMonthlyLimit.length() > 0 ? "checked" : "");
+            model.put("downloadLimit", downloadLimit);
+            model.put("uploadLimit", uploadLimit);
+            model.put("bandwidthMonthlyLimit", bandwidthMonthlyLimit);
+            model.put("storageLimit", storageLimit);
+            model.put("setPassword", setPassword.length() > 0 ? "checked" : "");
+
+            if (downloadLimitDetected == null || uploadLimitDetected == null)
+                detectBandwidthLimit();
```

Comment:
please always add curly braces around the if-body. This isn’t done everywhere in the existing code, but where we touch code, we should fix it.